pkg/agent/openflow/pipeline.go

```diff
@@ -2372,55 +2374,49 @@ func (f *featureService) serviceLBFlow(groupID binding.GroupIDType,
+		Cookie(f.cookieAllocator.Request(f.category).Raw()).
+		MatchProtocol(protocol).
+		MatchDstPort(svcPort, nil)
+
+	// EpToSelectRegMark is required to match the packets that haven't undergone Endpoint selection yet.
+	regMarksToMatch := []*binding.RegMark{EpToSelectRegMark}
+	// If "matchNodePortAddress" is true, which means that Service type is NodePort, "ToNodePortAddressRegMark" must be
+	// used to match packets. Otherwise, Service IP must be used to match packets.
+	if matchNodePortAddress {
+		regMarksToMatch = append(regMarksToMatch, ToNodePortAddressRegMark)
 	} else {
-		lbResultMark = EpSelectedRegMark
-	}
-	regMarksToLoad := []*binding.RegMark{lbResultMark, RewriteMACRegMark}
-	var flowBuilder binding.FlowBuilder
-	if serviceType == v1.ServiceTypeNodePort {
-		unionVal := (ToNodePortAddressRegMark.GetValue() << ServiceEPStateField.GetRange().Length()) + EpToSelectRegMark.GetValue()
-		// If externalTrafficPolicy of a NodePort Service is Cluster (nodeLocalExternal=false), it loads
-		// ToClusterServiceRegMark to indicate it. The mark will be checked later when determining if SNAT is required
-		// or not.
-		if !nodeLocalExternal {
-			regMarksToLoad = append(regMarksToLoad, ToClusterServiceRegMark)
-		}
-		flowBuilder = ServiceLBTable.ofTable.BuildFlow(priorityNormal).
-			Cookie(cookieID).
-			MatchProtocol(protocol).
-			MatchRegFieldWithValue(NodePortUnionField, unionVal).
-			MatchDstPort(svcPort, nil).
-			Action().LoadRegMark(regMarksToLoad...)
+		flowBuilder = flowBuilder.MatchDstIP(svcIP)
+	}
+
+	// RewriteMACRegMark is required to load for Service packets.
+	regMarksToLoad := []*binding.RegMark{RewriteMACRegMark}
+	// If SessionAffinity is set, the result of Endpoint selection should be cached by loading the EpToLearnRegMark.
+	// If SessionAffinity is not set, loading the EpSelectedRegMark indicates that the Endpoint selection is complete.
+	if withSessionAffinity {
+		regMarksToLoad = append(regMarksToLoad, EpToLearnRegMark)
 	} else {
-		// If externalTrafficPolicy of a LoadBalancer Service is Cluster (nodeLocalExternal=false), it loads
-		// ToClusterServiceRegMark to indicate it. The mark will be checked later when determining if SNAT is required
-		// or not.
-		if serviceType == v1.ServiceTypeLoadBalancer && !nodeLocalExternal {
-			regMarksToLoad = append(regMarksToLoad, ToClusterServiceRegMark)
-		}
-		flowBuilder = ServiceLBTable.ofTable.BuildFlow(priorityNormal).
-			Cookie(cookieID).
-			MatchProtocol(protocol).
-			MatchDstPort(svcPort, nil).
-			MatchDstIP(svcIP).
-			MatchRegMark(EpToSelectRegMark).
-			Action().LoadRegMark(regMarksToLoad...)
+		regMarksToLoad = append(regMarksToLoad, EpSelectedRegMark)
+	}
+	// If the Service is accessible externally, the ExternalAccessibleRegMark should be loaded to determine whether SNAT
+	// is required for the connection.
+	if externallyAccessible {
+		regMarksToLoad = append(regMarksToLoad, ExternalAccessibleRegMark)
```

Comment:
This is a bit misleading. It's not really about whether the Service is accessible externally but about whether it's the external destination of a Service being accessed.
Perhaps name the variable `externalAddress` and rename `matchNodePortAddress` to `nodePortAddress`.
```suggestion
	// If this is an external address of the Service, ToExternalAddressRegMark should be loaded to determine whether SNAT
	// is required for the connection later.
	if externalAddress {
		regMarksToLoad = append(regMarksToLoad, ToExternalAddressRegMark)
```

Reply:
Done

Reply:
Please check, I have updated this.